csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+struct Edge {
+  Expr* expr_ = nullptr;
+  size_t index_ = 0;
+
+  Edge(Expr* expr, size_t index) : expr_(expr), index_(index) {}
+
+  Val* val() const {
+    return expr_->input(index_);
+  }
+};
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `unaryOp(0) == 0`
+bool padCompatibleUnaryOp(UnaryOpType t) {
+  switch (t) {
+    case UnaryOpType::Cast:
+    case UnaryOpType::Abs:
+    case UnaryOpType::Asin:
+    case UnaryOpType::Asinh:
+    case UnaryOpType::Atan:
+    case UnaryOpType::Atanh:
+    case UnaryOpType::Ceil:
+    case UnaryOpType::Erf:
+    case UnaryOpType::Erfinv:
+    case UnaryOpType::Floor:
+    case UnaryOpType::Gelu:
+    case UnaryOpType::Imag:
+    case UnaryOpType::Silu:
+    case UnaryOpType::Neg:
+    case UnaryOpType::Real:
+    case UnaryOpType::Relu:
+    case UnaryOpType::Round:
+    case UnaryOpType::Sin:
+    case UnaryOpType::Sinh:
+    case UnaryOpType::Sqrt:
+    case UnaryOpType::Tan:
+    case UnaryOpType::Tanh:
+    case UnaryOpType::Trunc:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `binaryOp(0, x) == 0` & `binaryOp(x, 0) == 0`
+bool padCompatibleBinaryOp(BinaryOpType t) {
+  switch (t) {
+    case BinaryOpType::Add:
+    case BinaryOpType::Mul:
+    case BinaryOpType::Sub:
+    case BinaryOpType::BitwiseAnd:
+    case BinaryOpType::LogicalAnd:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// This function checks if we could propagate the PadOp on `val` to its
+// definition. i.e. changing the graph from
+//   inp --> definition() --> val --> PadOp --> padded_val
+// to
+//   inp --> PadOp --> padded_inp --> definition() --> padded_val
+//
+// returns true if `PadOp` should be propagated pass val->definition().
+//
+// NOTE 0: For details on the propagation logic, see Note [ PadOp Propagation
+// Rule ] NOTE 1: This function also update frontier accordingly, see note
+// below.
+bool shouldPropagatePad(Val* val, std::unordered_map<Val*, int64_t>& frontier) {
+  // NOTE [ Handling TV with Multiple Uses via Frontier ]
+  //
+  // Frontier counts the encounter of each Val during the propagation. When we
+  // reach a Val during the propagation, we add Val entry *temporarily* to the
+  // frontier map and increment its encounter count. We cannot propagate the
+  // PadOp over the producer of Val yet, because it still have other uses that
+  // expects the original Val. Until we see all uses of Val during the
+  // traversal, we know all uses can be replaced with the padded Val and hence
+  // it's safe to propagate `PadOp` across. We remove Val fron `frontier` map
+  // since we can push the replay to its producer. We return `true` to signal
+  // that further propagation is allowed.
+  if (val->isFusionOutput()) {
+    // skipping propagation when target is an output.
```

Comment:
Why?

Reply:
if val is a fusion output, I decided to stop the propagation here. Because after the propagation, we'll have `padded_val` instead, which means we have to preserve the old branch to have correct output `val`.

Note, this logic is moved to a new place in the refactor. I'll add a comment there.